zerver/tests/test_message_fetch.py

```diff
@@ -1197,6 +1197,91 @@ def test_successful_get_messages(self) -> None:
+                               status_code=401)
+
+    def test_unauthenticated_get_messages_with_web_public(self) -> None:
+        """
+        An unauthenticated call to GET /json/messages without valid
+        parameters in the `streams:web-public` narrow returns a 401.
+        """
+        post_params: Dict[str, Union[int, str, bool]] = {
+            "anchor": 1,
+            "num_before": 1,
+            "num_after": 1,
+            # "is:private" is not a is_web_public_compatible narrow.
+            "narrow": ujson.dumps([dict(operator="streams", operand="web-public"),
+                                   dict(operator="is", operand="private")]),
+        }
+        result = self.client_get("/json/messages", dict(post_params))
+        self.assert_json_error(result, 'Not logged in: API authentication or user session required',
+                               status_code=401)
+
+    def setup_web_public_test(self) -> None:
+        """
+        Send 3 messages, one in a web-public stream, one in a non web-public stream
+        and a private message.
+        """
+        user_profile = self.example_user('iago')
+        self.login('iago')
+        web_public_stream = self.make_stream('web-public-stream', is_web_public=True)
+        non_web_public_stream = self.make_stream('non-web-public-stream')
+        self.subscribe(user_profile, web_public_stream.name)
+        self.subscribe(user_profile, non_web_public_stream.name)
+
+        self.send_stream_message(user_profile, web_public_stream.name,
+                                 content="web-public message")
+        self.send_stream_message(user_profile, non_web_public_stream.name,
+                                 content="non web-public message")
+        self.send_personal_message(user_profile, self.example_user('hamlet'),
+                                   content="private message")
+        self.logout()
+
+    def test_get_messages_with_web_public(self) -> None:
+        """
+        An unauthenticated call to GET /json/messages with valid parameters
+        including `streams:web-public` narrow returns list of messages in the
+        `web-public` streams.
+        """
+        self.setup_web_public_test()
+
+        post_params = {
+            "anchor": 10000000000000000,
+            "num_before": 1,
+            "num_after": 1,
+            "narrow": ujson.dumps([dict(operator='streams', operand="web-public")]),
+        }
+        result = self.client_get("/json/messages", dict(post_params))
+
+        self.assert_json_success(result)
+        messages = ujson.loads(result.content)['messages']
+        # Out of last three messages, only 1 message is returned which is
+        # in a web-public stream. Other web-public messsages were discarded when
+        # when the query was limited to the last three messsages.
+        self.assert_length(messages, 1)
+        self.assertEqual(messages[0]['content'], '<p>web-public message</p>')
```

Comment:
Let's check a few more attributes of the returned message object in these tests: `sender_email` and `flags` at least, and maybe `avatar_url` too.  Also the ID, which is returned by `self.send_stream_message`, is a good idea too.

Reply:
Okay.